packages/flutter/lib/src/widgets/focus_scope.dart

```diff
@@ -1,145 +1,369 @@
+    properties.add(DiagnosticsProperty<FocusScopeNode>('node', focusNode, defaultValue: null));
   }
 
   @override
-  _FocusScopeState createState() => _FocusScopeState();
+  _FocusState createState() => _FocusState();
 }
 
-class _FocusScopeState extends State<FocusScope> {
+class _FocusState extends State<Focus> {
+  FocusNode _internalNode;
+  FocusNode get node => widget.focusNode ?? _internalNode;
+  bool _hasFocus;
   bool _didAutofocus = false;
 
-  @override
-  void didChangeDependencies() {
-    super.didChangeDependencies();
-    if (!_didAutofocus && widget.autofocus) {
-      FocusScope.of(context).setFirstFocus(widget.node);
-      _didAutofocus = true;
+  FocusNode _createNode() {
+    return FocusNode(
+      debugLabel: widget.debugLabel,
+    );
+  }
+
+  void _initNode() {
+    if (widget.focusNode == null) {
+      // Only create a new node if the widget doesn't have one.
+      _internalNode ??= _createNode();
     }
+    _hasFocus = node.hasFocus;
+    // Add listener even if the _internalNode existed before, since it should
+    // not be listening already if we're re-using a previous one.
+    node.addListener(_handleFocusChanged);
+    node.onKey = widget.onKey;
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _initNode();
   }
 
   @override
   void dispose() {
-    widget.node.detach();
+    // Regardless of the node owner, we need to remove it from the tree and stop
+    // listening to it.
+    node.removeListener(_handleFocusChanged);
+    node.onKey = null;
+    node.detach();
+    // Don't manage the lifetime of external nodes given to the widget, just the
+    // internal node.
+    _internalNode?.dispose();
     super.dispose();
   }
 
+  void _handleFocusChanged() {
+    if (!mounted) {
+      return;
+    }
+    if (_hasFocus != node.hasFocus) {
+      setState(() {
+        _hasFocus = node.hasFocus;
+      });
+      if (widget.onFocusChange != null) {
+        widget.onFocusChange(node.hasFocus);
+      }
+    }
+  }
+
+  @override
+  void didUpdateWidget(Focus oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.focusNode == widget.focusNode) {
+      // Nothing we care about changed.
+      return;
+    }
+    if (oldWidget.focusNode == null && widget.focusNode != null) {
+      // We're no longer using the node we were managing. We don't stop managing
+      // it until dispose, so do nothing yet: we might re-use it eventually, and
+      // calling dispose on it here will confuse other widgets that haven't yet
+      // been notified of a widget change and might still be listening.
+      _internalNode?.removeListener(_handleFocusChanged);
+      widget.focusNode?.addListener(_handleFocusChanged);
+    } else if (oldWidget.focusNode != null && widget.focusNode == null) {
+      oldWidget.focusNode?.removeListener(_handleFocusChanged);
+      // We stopped using the external node, and now we need to manage one.
+      _initNode();
```

Comment:
i think i'm convinced it's correct, but the lack of symmetry here (remove here, but add is in initNode) makes me nervous.

Reply:
Yeah, I don't really like this either.